### PR TITLE
Rule migrations

### LINKED ISSUE #7160: Migrate TransformUncorrelatedScalarToJoin to iterative optimizer

### PR DESCRIPTION
**Pull request by ArturGajowy**
Fixes #7160, #7159 and #7158

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 1]
```diff
@@ -57,7 +57,6 @@
 
 import static com.facebook.presto.sql.analyzer.TypeSignatureProvider.fromTypeSignatures;
 import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
-import static com.facebook.presto.sql.planner.optimizations.Predicates.isInstanceOfAny;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: unrelated

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformUncorrelatedScalarToJoin.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                .filter(applyNode -> canBeReplacedByScalarJoin(applyNode, lookup))
+                .map(applyNode -> createReplacementJoinNode(applyNode, idAllocator, lookup));
+    }
+
+    private boolean canBeReplacedByScalarJoin(ApplyNode applyNode, Lookup lookup)
+    {
+        return applyNode.getCorrelation().isEmpty() && isResolvedScalarSubquery(applyNode, lookup);
+    }
+
+    private PlanNode createReplacementJoinNode(ApplyNode applyNode, PlanNodeIdAllocator idAllocator, Lookup lookup)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformUncorrelatedScalarToJoin.java]: any changes here?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java [hunk 3]
```diff
@@ -164,7 +164,7 @@ private PlanNode rewriteQuantifiedApplyNode(ApplyNode node, QuantifiedComparison
                     node.getId(),
                     context.rewrite(node.getInput()),
                     subqueryPlan,
-                    Assignments.identity(minValue, maxValue),
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: typo in commit message (quantified)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 4]
```diff
@@ -150,7 +150,10 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                 new MergeProjections(),
                 new TransformExistsApplyToScalarApply(metadata),
                 new TransformQuantifiedComparisonApplyToScalarApply(metadata),
-                new RemoveUnreferencedScalarInputApplyNodes(),
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: change commit message to 
```
Migrate RemoveUnreferencedScalarInputApplyNodes to iterative optimizer
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveUnreferencedScalarInputApplyNodes.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import static com.facebook.presto.sql.planner.optimizations.ScalarQueryUtil.isScalar;
+import static com.facebook.presto.util.Optionals.tryCast;
+
+/**
+ * Remove resolved ApplyNodes with unreferenced scalar input, e.g: "SELECT (SELECT 1)".
+ */
+public class RemoveUnreferencedScalarInputApplyNodes
+        implements Rule
+{
+
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveUnreferencedScalarInputApplyNodes.java]: Blank line after `{`.
It would not pass checkstyle I believe.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveUnreferencedScalarInputApplyNodes.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public Optional<PlanNode> apply(PlanNode planNode, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+    {
+        return tryCast(planNode, ApplyNode.class)
+                .filter(applyNode -> canBeReplacedByItsSubquery(applyNode, lookup))
+                .map(applyNode -> lookup.resolve(applyNode.getSubquery()));
+    }
+
+    public boolean canBeReplacedByItsSubquery(ApplyNode applyNode, Lookup lookup)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveUnreferencedScalarInputApplyNodes.java]: Maybe drop `Its`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformUncorrelatedScalarToJoin.java (hunk 2 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformUncorrelatedScalarToJoin.java]: Not really - just extracted two variables and resolved them

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java [hunk 7]
```diff
@@ -1213,7 +1214,7 @@ private void printAssignments(Assignments assignments, int indent)
 
         private String formatOutputs(Iterable<Symbol> symbols)
         {
-            return Joiner.on(", ").join(Iterables.transform(symbols, input -> input + ":" + types.get(input).getDisplayName()));
+            return Joiner.on(", ").join(Iterables.transform(symbols, input -> input + ":" + types.getOrDefault(input, UNKNOWN).getDisplayName()));
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: +1 Investigate whether this is not a bug somewhere else in the code that the type is missing for symbol. I wouldn't assume that is is `UNKNOWN` by default.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java [hunk 8]
```diff
@@ -30,33 +30,26 @@
...
-
-        Optional<List<Symbol>> dependencies = pruneInputs(child.getOutputSymbols(), parent.getAssignments().getExpressions());
-        if (!dependencies.isPresent()) {
-            return Optional.empty();
-        }
-
-        List<Symbol> newOutputs = dependencies.get();
+        return tryCast(node, ProjectNode.class)
+            .flatMap((ProjectNode parent) -> lookup.resolve(parent.getSource(), ValuesNode.class)
+            .flatMap((ValuesNode child) -> pruneInputs(child.getOutputSymbols(), parent.getAssignments().getExpressions())
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java]: use proper alignment here. It's not obvious (from formatting) why you have `parent` instance in this line.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneValuesColumns.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+public class TestPruneValuesColumns
+{
+    private final RuleTester tester = new RuleTester();
+
+    @Test
+    public void test()
+            throws Exception
+    {
+        tester.assertThat(new PruneValuesColumns())
+                .on(p ->
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneValuesColumns.java]: don't use one letters variable names. What is a `p`, projection, predicate?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughMarkDistinct.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+public class TestPushLimitThroughMarkDistinct
+{
+    private final RuleTester tester = new RuleTester();
+
+    @Test
+    public void test()
+            throws Exception
+    {
+        tester.assertThat(new PushLimitThroughMarkDistinct())
+                .on(p ->
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughMarkDistinct.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java [hunk 11]
```diff
@@ -164,7 +164,7 @@ private PlanNode rewriteQuantifiedApplyNode(ApplyNode node, QuantifiedComparison
...
+                    Assignments.identity(minValue, maxValue, countAllValue, countNonNullValue),
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: make it a separate PR?

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: Add test case that it fixes

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java (hunk 7 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: This happened when running a test that was not well-defined yet. Nevertheless, PlanPrinter should be robust enough to cope with ill-defined trees. If there is a problem with missing types in production code, it'll get caught by the code using them anyway.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java [hunk 12]
```diff
@@ -13,30 +13,64 @@
...
 public final class ScalarQueryUtil
 {
     private ScalarQueryUtil() {}
 
+    /**
+     * @return true when subquery is scalar and its output symbols are directly mapped to the ApplyNode's output symbols
+     */
+    public static boolean isResolvedScalarSubquery(ApplyNode applyNode)
+    {
+        return isScalar(applyNode.getSubquery(), x -> x) && applyNode.getSubqueryAssignments().getExpressions().stream()
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: This is to remain consistence with other code using lookup functions

File: presto-main/src/main/java/com/facebook/presto/util/Optionals.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class Optionals
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/util/Optionals.java]: I wouldn't say it matters at all

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java [hunk 14]
```diff
@@ -23,43 +23,39 @@
...
+                                p.getCount(),
+                                false,
+                                d.getHashSymbol())));
     }
 
     /**
      * Whether this node corresponds to a DISTINCT operation in SQL
      */
-    private boolean isDistinct(AggregationNode node)
+    private static boolean isDistinct(AggregationNode node)
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: I think you have a point here, but this is not the scope of this PR. A quick search revealed 3 occurrences of similar code throughout the codebase. Feel free to open an issue for unifying those.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneValuesColumns.java (hunk 9 above)
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneValuesColumns.java]: I don't think it's pracitcal in this case. 'p' stands for 'PlanBuilder'. Please see the original https://github.com/prestodb/presto/commit/74165ea commit

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughMarkDistinct.java (hunk 10 above)
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughMarkDistinct.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java (hunk 11 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: The tests in the following commits break without that. So it needs to stay in this PR and there are tests for that. Plus, it's only needed due to the requirements IterativeOptimizer introduces

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/BasePlanTest.java [hunk 15]
```diff
@@ -46,6 +47,7 @@ public BasePlanTest(Map<String, String> sessionProperties)
         Session.SessionBuilder sessionBuilder = testSessionBuilder()
                 .setCatalog("local")
                 .setSchema("tiny")
+                .setSystemProperty(ITERATIVE_OPTIMIZER, "true")
```
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/BasePlanTest.java]: Not in BasePlanTest

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java (hunk 8 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java]: I'd call *this* the proper alignment for chaining a bunch of flatmaps. The nesting here is not really relevant and can in fact decrease readability even more than the surprising availability of parent in the scope where the second flatmap is evaluated.

**Comment by ArturGajowy** [comment]
Comments applied.

@martint please have a look. You mentioned chaining `flatMap`s can get unwieldy. Are they unwieldy in this PR? Should we stay with them in this PR since they improve conciseness & readability? Or should I drop the `Refactor * to leverage Optional's methods` and related commits, and rewrite the migrated rules to ifs + casts?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java (hunk 12 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: `x -> x` is not needed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java [hunk 16]
```diff
@@ -13,30 +13,64 @@
...
+                .allMatch(expression -> expression instanceof SymbolReference);
+    }
+
+    public static boolean isResolvedScalarSubquery(ApplyNode applyNode, Lookup lookup)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: replace `Lookup` with `Function` here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java [hunk 17]
```diff
@@ -13,30 +13,64 @@
...
+
+    public static boolean isScalar(PlanNode node, Function<PlanNode, PlanNode> lookupFunction)
+    {
+        return node.accept(new IsScalarPlanVisitor(lookupFunction), null);
     }
 
     private static final class IsScalarPlanVisitor
             extends PlanVisitor<Void, Boolean>
     {
+        private final Function<PlanNode, PlanNode> lookupFunction;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: please add a comment why it is needed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestQuantifiedComparison.java [hunk 18]
```diff
@@ -96,8 +96,7 @@ private void assertQuantifiedComparison(String query)
         assertPlan(query, anyTree(
                 node(JoinNode.class,
                         tableScan("orders"),
-                        anyTree(
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestQuantifiedComparison.java]: I think one should finally fix this, to make `anyTree` accept empty tree. Would you accept this honor?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Lookup.java [hunk 19]
```diff
@@ -15,6 +15,10 @@
 
 import com.facebook.presto.sql.planner.plan.PlanNode;
 
+import java.util.Optional;
+
+import static com.facebook.presto.util.Optionals.tryCast;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Lookup.java]: `tryCast` should belong to `Types`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java (hunk 7 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: I would prefer to fix the test, or test facility to use well-defined plans.

File: presto-main/src/main/java/com/facebook/presto/util/Optionals.java (hunk 13 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/Optionals.java]: Rename it to `Types` and make sure this commit is added before any other which is using this.

+1 for final

File: presto-main/src/test/java/com/facebook/presto/util/TestOptionals.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+package com.facebook.presto.util;
+
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static com.facebook.presto.util.Optionals.tryCast;
+import static org.testng.Assert.assertEquals;
+
+public class TestOptionals
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/util/TestOptionals.java]: What a nice tests!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java (hunk 11 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: it is not that this optimizer introduced that. It is rather a bug in previous optimizer.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java (hunk 7 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: So if the test is now defined, you can switch back to:
```
types.get(input)
```

Assumption that when there is no type in map, then the type is `UNKNOWN` is incorrect. There is difference between no type computed and computed `UNKNOWN` type

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java (hunk 12 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: Other code using lookup functions use `identity()`. See `isScalar` in the same class

File: presto-main/src/main/java/com/facebook/presto/util/Optionals.java (hunk 13 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/util/Optionals.java]: Convention is to use `final` for utility classes in presto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java (hunk 14 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: If our goal is not only to migrate rules, but also refactor the code so it's more readable (less error prone), then this is in the scope of this PR. 3 occurrences seems enough to make the function part of `AgggregationNode`, but it's up to you to decide

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java (hunk 8 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java]: I think it's not proper because it doesn't visually explain variables visibility. Additionally, it's different than alignment in other Presto code (breaks convention). This also means that it doesn't work with automatic code formatter, which is a no-go.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestQuantifiedComparison.java [hunk 21]
```diff
@@ -96,8 +96,7 @@ private void assertQuantifiedComparison(String query)
...
-                                node(AggregationNode.class,
-                                        node(ValuesNode.class))))));
+                        node(AggregationNode.class,
```
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestQuantifiedComparison.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 1 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: fixed, I think

**Comment by ArturGajowy** [comment]
Applied comments and rebased. @kokosing and @sopel39 please have another look

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestQuantifiedComparison.java (hunk 18 above)
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestQuantifiedComparison.java]: I'm not sure about the complexity of this task. For sure I wouldn't like to do this in this PR

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java (hunk 8 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java]: I stand corrected :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java [hunk 22]
```diff
@@ -30,33 +30,26 @@
...
-
-        List<Symbol> newOutputs = dependencies.get();
+        return tryCast(node, ProjectNode.class)
+                .flatMap((ProjectNode parent) -> lookup.resolve(parent.getSource(), ValuesNode.class)
+                        .flatMap((ValuesNode child) -> pruneInputs(child.getOutputSymbols(), parent.getAssignments().getExpressions())
+                                .map((List<Symbol> dependencies) ->
+                                        new ProjectNode(
+                                                parent.getId(),
+                                                createNewValuesNode(child, dependencies),
+                                                parent.getAssignments()))));
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java]: +1

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java [hunk 23]
```diff
@@ -30,33 +30,26 @@
...
-        ValuesNode values = (ValuesNode) child;
-
-        Optional<List<Symbol>> dependencies = pruneInputs(child.getOutputSymbols(), parent.getAssignments().getExpressions());
-        if (!dependencies.isPresent()) {
-            return Optional.empty();
-        }
-
-        List<Symbol> newOutputs = dependencies.get();
+        return tryCast(node, ProjectNode.class)
+                .flatMap((ProjectNode parent) -> lookup.resolve(parent.getSource(), ValuesNode.class)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneValuesColumns.java]: do you need explicit types in lambda arguments? Without them the lambda will be more compact

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java (hunk 11 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: Include tests that this change fixes as part of the commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java (hunk 7 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: What was the problem @ArturGajowy ?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java (hunk 11 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: They're in the following `Migrate TransformUncorrelatedScalarToJoin to Rule` commit, yet this change IMO deserves a separate commit. OTOH having a separate test in this commit asserting that some plan node has these specific assignments would be testing implementation details. The functionality that this code was introduced for in #6960 is thoroughly tested on query results level.

**Comment by ArturGajowy** [comment]
@martint kindly bump

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java [hunk 24]
```diff
@@ -14,29 +14,69 @@
...
             extends PlanVisitor<Void, Boolean>
     {
+        /*
+        With the iterative optimizer, plan nodes are replaced with
+        GroupReference nodes. This requires the rules that look deeper
+        in the tree than the rewritten node and its immediate sources
+        to use Lookup for resolving the nodes corresponding to the
+        GroupReferences.
+        */
+        private final Function<PlanNode, PlanNode> lookupFunction;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: What about using #7686 instead?

Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: I'd like to avoid depending on Lookup itself, but the line `return isScalar(node, identity());` several lines above could then be replaced with `return isScalar(node, noLookup()::lookup);` to make things more clear :D

**Comment by martint** [comment]
I'm going to start reviewing this. Can you rebase?

**Comment by ArturGajowy** [comment]
Superseded by #7294, closing. cc @martint
